File: F#/chap06/Chap0604.fs
namespace Chap0604

type EmailAddress = Undefiend

type VerifiedEmailAddress = Undefiend

module C0604a =
    type CustomerEmail =
        { EmailAddress: EmailAddress
          IsVerified: bool }

module C0604b =
    type CustomerEmail =
        | Unverified of EmailAddress
        | Verified of EmailAddress

module C0604c =
    type CustomerEmail =
        | Unverified of EmailAddress
        | Verified of VerifiedEmailAddress

type Name = Undefiend
type EmailContactInfo = Undefiend
type PostalContactInfo = Undefiend


module C0604d =
    type Contact =
        { Name: Name
          Email: EmailContactInfo
          Address: PostalContactInfo }

module C0604e =
    type Contact =
        { Name: Name
          Email: EmailContactInfo option
          Address: PostalContactInfo option }

module C0604f =
    type BothContactMethods =
        { Email: EmailContactInfo
          Address: PostalContactInfo }

    type ContactInfo =
        | EmailOnly of EmailContactInfo
        | AddrOnly of PostalContactInfo
        | EmailAndAddr of BothContactMethods

    type Contact =
        { Name: Name; ContactInfo: ContactInfo }

File: Scala/chap06/src/main/scala/Chap0604.scala
import Chap0604.*

object Chap0604:
  case class EmailAddress(value: String)

  case class VerifiedEmailAddress(value: String)

  case class Name(value: String)

  case class EmailContactInfo(value: String)

  case class PostalContactInfo(value: String)

object C0604a:
  case class CustomerEmail(
                            email: EmailAddress,
                            isVerified: Boolean,
                          )

object C0604b:
  enum CustomerEmail:
    case Unverified(email: EmailAddress)
    case Verified(email: EmailAddress)


object C0604c:
  enum CustomerEmail:
    case Unverified(email: EmailAddress)
    case Verified(email: VerifiedEmailAddress)


object C0606d:
  case class Contact(
                      name: Name,
                      email: EmailContactInfo,
                      address: PostalContactInfo,
                    )

object C0606e:
  case class Contact(
                      name: Name,
                      email: Option[EmailContactInfo],
                      address: Option[PostalContactInfo],
                    )

object C0606f:
  case class BothContactMethods(
                                 email: EmailContactInfo,
                                 address: PostalContactInfo,
                               )

  enum ContactInfo:
    case Email(email: EmailContactInfo)
    case Address(address: PostalContactInfo)
    case EmailAndAddr(email: EmailContactInfo, address: PostalContactInfo)

  case class Contact(
                      name: Name,
                      info: ContactInfo,
                    )

object C060401:
  case class UnvalidatedAddress(value: String)

  case class ValidatedAddress(value: String)

  type AddressValidationService = UnvalidatedAddress => Option[ValidatedAddress]

  case class UnvalidatedOrder(shippingAddress: UnvalidatedAddress)

  case class ValidatedOrder(shippingAddress: ValidatedAddress)
